### DISCUSSION TITLE
A small collection of "How To" worked examples

### DISCUSSION THREAD

**Discussion by chrisgoringe** [chrisgoringe/cg-use-everywhere#171] [Ideas]
I wonder if we might be able to build a small collection of "How To"s - that illustrate how to do some clever things like @JorgeR81's [group bypasses](https://github.com/chrisgoringe/cg-use-everywhere/discussions/168#discussioncomment-8871778) 

I'll create an section in the Discussions for it - maybe a single thread for each "How To" - for anyone who feels like contributing!

**Comment by JorgeR81** [comment]
I will share a version of that workflow once it's completed. 

One of the ideas behind that workflow is that each bypassable ( muted ) group, could also be saved as a template.
So I can start with a small workflow, and build up from that.
I could load a new node template, and it would just connect automatically !

The [ aegisflow_utility_nodes ] does Auto Grouping of Nodes placed as templates.
So no other action would be needed.
 
https://github.com/aegis72/aegisflow_utility_nodes

The node suite creator also has this workflow system for Comfy UI, based on UE? nodes + node templates. 
I haven't tried yet, but it's probably the one most evolved workflow systems for Comfy UI. 

https://www.youtube.com/@MajorStudioArt/videos

**Comment by JorgeR81** [comment]
I was using UE links mostly between groups, as described above.
But I'm discovering that they just are as useful for links inside groups. 
The group and color restrictions are extremely useful.

**UE can make all the _"back end"_ connections automatically !**
This saves time and prevents errors making all these tedious connections

And I think I will not need switch nodes anymore !
I can now use regular wires to make my selections.  
These allow a better visualization of my selection, since there are no other visibe wires.

Here is a preprocessor group, sending images to 2 ControlNet groups. 
The UE nodes with the green circles are using "group restriction". 

![sw4](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/5607327a-68ca-4e29-9b14-478a0223b6db)

Here are all the "backend" connections, via UE.
I can send the same image and the same "resolution" setting to all preprocessors.


![sw5](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/e04ba86c-60a9-4d3f-8d34-39dd0787e025)


Next is a more complex use case, with group restriction and color restriction.

This is a group where the main prompt is combined with a new prompt. 
We can do many operations, even 2nd order ones, e.g.: _combine [ average + concat ]_.
I tried this before, without UE.
If you do this manually, it is really easy to make the wrong connections. 

I also used a switch node to make my selection, because I didn't even want to touch cables in that group, in order not to risk messing up my "backend "configuration. 

But now, I just can use regular wires to "switch" my selection. Since they are the only ones visible, they are very easy and fun to use. 

![sw2](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/b69addca-4cc2-4d8b-b488-2c2b427b79b8)

Here are all the "back end" conections, with UE wires.

![sw1](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/9af5337c-0e74-44e3-a7a3-6b53161bb904)

**Comment by JorgeR81** [comment]
Once we have created the "back end" connections ( to add all the options ), we can just add more UE? nodes to the "front end" ( to have extra "switches", with more selected options ).

Here I'm still sending the combined prompt to the "Sampler 1" group ( via the "ControlNet" groups sequence ).

But for the "Sampler 2" group ( Refiner ), I'm sending only the 2nd prompt ( with with priority boost ).

In my case, the "ControlNet1" and "ControlNet2" groups, have priority 1 and 2 already, and they send to both "Sampler" groups.

So, for this to work, with the ControlNet groups enabled ( unmuted ), I need to send to the "Sampler 2" group with priority 3.

![cp](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/f159a18d-6f7c-4df0-bd5e-985d3688973f)

**Comment by JorgeR81** [comment]
Using Fast Group Muter ( rgthree ) nodes, we can create a "Control Panel'', for all the group sequences, for each type of data. 

Here I have sequences for conditioning, and for the image ( after being created by the KSampler ).
I just need to toggle the groups I want to enable in the sequence. 

If I enable the ControlNet Preprocessor group here, the option to be used, will be the one selected with the regular wires, inside the group, as shown in the post above. 

I can use the arrows in the rgthree node, to go quickly navigate to any group, and change my options.
 An with the Bookmark node ( rgthree ), I can go back to the "Control Panel", via keybind.

Next I'm planning to add another sequence, for "model" ( purple wires ), with IPAdapter groups.

![vn](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/473be6bc-157c-4c86-9d0a-adbce3cbb1a4)

**Comment by JorgeR81** [comment]
Here is how I'm organizing the groups in the workflow.

I'm using the same seed, sampler and scheduler, for all the stages, via UE? nodes, and KSampler Config ( rgthree ) nodes. 

![rg3](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/66c4cc38-d443-4c9f-850b-9d1ec5c67f23)

To keep a clean look, I'm merging nodes.
I can group the Sampler with the Image preview, and I also have a global group for the final preview. 
I keep UE? nodes separe and on top, like an "address bar".

![wf0](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/03a3e301-bb49-409a-8613-be2f53267fa7)

In this case, Face Detailer ( before Ultimate Upscaler ) is removing the paint from the face, ( despite being in the prompt ).
So I'll just disable Face Detailer, in the "Control Panel", since the face also looks good without it.

![z1](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/c0552273-61f1-4898-a159-88bcc4c1d671)

![wf2](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/39ae98be-1972-4bb1-81c0-baa94a31e8e7)
